From 4d610b2e2a965f7cb87b9e698cd2384135958c42 Mon Sep 17 00:00:00 2001
From: Josef Wegner <josef.wegner@outlook.com>
Date: Tue, 15 Sep 2026 20:20:43 +0200
Subject: [PATCH 3/3] Added better handling of Amiga file paths: - Use SDK: and
 GCC: instead of /SDK/ and /GCC/ - Some fixes to the newlib, clib2 and clib4
 specs - Prevent gcc to add a directory separator after drive separator -
 Added xm-amigaos.h for gcc to understand Amiga file paths better

---
 gcc/config.host                               |  2 +-
 gcc/config/rs6000/amigaos.h                   | 13 +++++----
 .../xm-mingw32.h => rs6000/xm-amigaos.h}      | 29 +++++++++----------
 libiberty/make-temp-file.c                    |  6 +++-
 4 files changed, 28 insertions(+), 22 deletions(-)
 copy gcc/config/{i386/xm-mingw32.h => rs6000/xm-amigaos.h} (52%)

diff --git a/gcc/config.host b/gcc/config.host
index 2a11c603aad2857322b391d9c5d8638a3047de0c..cc73de723d6f585eb8ba33c013c44a54ef2c05fe 100644
--- a/gcc/config.host
+++ b/gcc/config.host
@@ -253,13 +253,13 @@ case ${host} in
     host_xmake_file="${host_xmake_file} i386/x-darwin"
     ;;
   powerpc-*-amigaos*) # AmigaOS 4
     prefix=/gcc
     local_prefix=/gcc
     host_can_use_collect2=no
-    host_xm_defines=HOST_LACKS_INODE_NUMBERS
+    host_xm_file=rs6000/xm-amigaos.h
     host_xmake_file="${host_xmake_file} rs6000/x-amigaos"
     ;;
   powerpc-*-darwin*)
     out_host_hook_obj="${out_host_hook_obj} host-ppc-darwin.o"
     host_xmake_file="${host_xmake_file} rs6000/x-darwin"
     ;;
diff --git a/gcc/config/rs6000/amigaos.h b/gcc/config/rs6000/amigaos.h
index 802d22903609042c3aef95233eb2344ef135b368..e0e635b88f741ad06420bc752ee55eb9ede125a6 100644
--- a/gcc/config/rs6000/amigaos.h
+++ b/gcc/config/rs6000/amigaos.h
@@ -171,14 +171,17 @@ ASM_OPT_ANY
 
 /*#define STANDARD_INCLUDE_DIR "/GCC/include"*/
 /*#undef SYSTEM_INCLUDE_DIR *//* So that the include path order is the same in native and cross compilers */
 #undef LOCAL_INCLUDE_DIR
 
 #ifndef CROSS_DIRECTORY_STRUCTURE
-#define BASE_GCC_SPEC "/GCC/"
-#define BASE_SDK_SPEC "/SDK/"
+/* Native AmigaOS installations use assigns rather than Unix-root paths.
+   In particular, /SDK/ means SDK on the current volume, while SDK: refers
+   to the SDK assign used by the system and development environments.  */
+#define BASE_GCC_SPEC "GCC:"
+#define BASE_SDK_SPEC "SDK:"
 #else
 #define BASE_GCC_SPEC EXEC_PREFIX
 #define BASE_SDK_SPEC EXEC_PREFIX "ppc-amigaos/SDK/"
 #endif
 
 #define LIB_SUBDIR_TYPE_SPEC "\
@@ -212,13 +215,13 @@ ASM_OPT_ANY
 -idirafter %(base_sdk)clib2/include -idirafter %(base_sdk)local/clib2/include"
 
 #define LIB_SUBDIR_CLIB2_SPEC "%{mcrt=clib2-ts:lib.threadsafe; :lib}%(lib_subdir_type)"
 
 #define LINK_CLIB2_SPEC "\
 -L%(base_sdk)clib2/%(lib_subdir_clib2) \
--L%(base_gcc)lib/gcc/ppc-amigaos/%(version)/%{mcrt=clib2-ts:clib2-ts; :clib2}/lib%(lib_subdir_type) \
+-L%(base_gcc)lib/gcc/ppc-amigaos/%(version)/%{mcrt=clib2-ts:clib2-ts; :clib2}%(lib_subdir_type) \
 -L%(base_sdk)local/clib2/%(lib_subdir_clib2)"
 
 #define STARTFILE_CLIB2_SPEC "\
 %(base_sdk)clib2/%{mcrt=clib2-ts:lib.threadsafe; :lib}" \
                  "%{!msoft-float:%(lib_subdir_type)}/crtbegin.o \
 %(base_sdk)clib2/%{mcrt=clib2-ts:lib.threadsafe; :lib}" \
@@ -234,13 +237,13 @@ ASM_OPT_ANY
 -idirafter %(base_sdk)clib4/include -idirafter %(base_sdk)local/clib4/include"
 
 #define LIB_SUBDIR_CLIB4_SPEC "lib%(lib_subdir_type)"
 
 #define LINK_CLIB4_SPEC "\
 -L%(base_sdk)clib4/%(lib_subdir_clib4) \
--L%(base_gcc)lib/gcc/ppc-amigaos/%(version)/clib4/lib%(lib_subdir_type) \
+-L%(base_gcc)lib/gcc/ppc-amigaos/%(version)/clib4%(lib_subdir_type) \
 -L%(base_sdk)local/clib4/%(lib_subdir_clib4)"
 
 #define STARTFILE_CLIB4_SPEC "\
 %{shared: %(base_sdk)clib4/%(lib_subdir_clib4)/shcrtbegin.o} %{!shared: %(base_sdk)clib4/%(lib_subdir_clib4)/crtbegin.o} %{!shared: %(base_sdk)clib4/%(lib_subdir_clib4)/crt0.o}"
 
 #define ENDFILE_CLIB4_SPEC "\
@@ -283,13 +286,13 @@ ASM_OPT_ANY
 -idirafter %(base_sdk)newlib/include -idirafter %(base_sdk)local/newlib/include"
 
 #define LIB_SUBDIR_NEWLIB_SPEC "lib%(lib_subdir_type)"
 
 #define LINK_NEWLIB_SPEC "\
 -L%(base_sdk)newlib/%(lib_subdir_newlib) \
--L%(base_gcc)lib/gcc/ppc-amigaos/%(version)/newlib/%(lib_subdir_newlib) \
+-L%(base_gcc)lib/gcc/ppc-amigaos/%(version)/newlib%(lib_subdir_type) \
 -L%(base_sdk)local/newlib/%(lib_subdir_newlib)"
 
 /* newlib startfile should work for all library flavours */
 #define STARTFILE_NEWLIB_SPEC "\
 %{shared: %(base_sdk)newlib/%(lib_subdir_newlib)/shcrtbegin.o} %{!shared: %(base_sdk)newlib/%(lib_subdir_newlib)/crtbegin.o}"
 
diff --git a/gcc/config/i386/xm-mingw32.h b/gcc/config/rs6000/xm-amigaos.h
similarity index 52%
copy from gcc/config/i386/xm-mingw32.h
copy to gcc/config/rs6000/xm-amigaos.h
index c005449bce7dee0c6395ddb37a79b858e101f916..ce3ab3b9a4da5c1529b4aae85d3045d7f3dd8029 100644
--- a/gcc/config/i386/xm-mingw32.h
+++ b/gcc/config/rs6000/xm-amigaos.h
@@ -1,9 +1,8 @@
-/* Configuration for GCC for hosting on Windows32.
-   using GNU tools and the Windows32 API Library.
-   Copyright (C) 1997-2023 Free Software Foundation, Inc.
+/* Configuration for GCC when hosted on AmigaOS 4.
+   Copyright (C) 2026 Free Software Foundation, Inc.
 
 This file is part of GCC.
 
 GCC is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free
 Software Foundation; either version 3, or (at your option) any later
@@ -15,25 +14,25 @@ FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 for more details.
 
 You should have received a copy of the GNU General Public License
 along with GCC; see the file COPYING3.  If not see
 <http://www.gnu.org/licenses/>.  */
 
-#define HOST_EXECUTABLE_SUFFIX ".exe"
-
+/* AmigaOS uses ':' for volume and assign names, so it cannot use ':' as
+   the separator in PATH-like variables.  */
 #undef PATH_SEPARATOR
 #define PATH_SEPARATOR ';'
 
-/* This is the name of the null device on windows.  */
-#define HOST_BIT_BUCKET "nul"
+/* AmigaOS uses NIL: as its null device.  */
+#define HOST_BIT_BUCKET "NIL:"
 
-/*  The st_ino field of struct stat is always 0.  */
+/* AmigaOS has no useful inode numbers.  */
 #define HOST_LACKS_INODE_NUMBERS
 
-#ifdef __MINGW32__
-#undef __USE_MINGW_ANSI_STDIO
-#define __USE_MINGW_ANSI_STDIO 1
-#else
-/* MSVCRT does not support the "ll" format specifier for printing
-   "long long" values.  Instead, we use "I64".  */
-#define HOST_LONG_LONG_FORMAT "I64"
+/* Make the common filename helpers understand Amiga assigns and
+   case-insensitive filenames.  '/' remains the directory separator.  */
+#ifndef HAVE_DOS_BASED_FILE_SYSTEM
+# define HAVE_DOS_BASED_FILE_SYSTEM 1
+#endif
+#ifndef HAVE_CASE_INSENSITIVE_FILE_SYSTEM
+# define HAVE_CASE_INSENSITIVE_FILE_SYSTEM 1
 #endif
diff --git a/libiberty/make-temp-file.c b/libiberty/make-temp-file.c
index dfb7b916ec56cad63081e3609730e59b8cbff18b..14dd881c3a0f6fceea9b4999226921999c83c187 100644
--- a/libiberty/make-temp-file.c
+++ b/libiberty/make-temp-file.c
@@ -155,13 +155,17 @@ choose_tmpdir (void)
 	base = ".";
       /* Append DIR_SEPARATOR to the directory we've chosen
 	 and return it (skip if the path already ends with a separator).  */
       len = strlen (base);
       tmpdir = XNEWVEC (char, len + 2);
       strcpy (tmpdir, base);
-      if (!IS_DIR_SEPARATOR (tmpdir[len - 1]))
+      if (!IS_DIR_SEPARATOR (tmpdir[len - 1])
+#ifdef __amigaos__
+	  && tmpdir[len - 1] != ':'
+#endif
+	  )
 	{
 	  tmpdir[len] = DIR_SEPARATOR;
 	  tmpdir[len + 1] = '\0';
 	}
       memoized_tmpdir = tmpdir;
 #else /* defined(_WIN32) && !defined(__CYGWIN__) */
-- 
2.47.3

